### PR TITLE
Add append and copy

### PR DESCRIPTION
**Pull request by jcp19**
This PR adds support for the `append` and `copy` built-in methods. I made some changes to the way we handle `BuiltInTags`. In particular, I introduced a file `BuiltInMemberTyping` that handles the well-definedness checks and typing of `BuiltInTags`. The reason for this change is that I need the type information (for checking assignability) to correctly type check functions `copy` and `append`.

I picked the following specs for `append` and `copy`
```go
requires p > 0
requires forall i int :: { dst[i] } 0 <= i && i < len(dst) ==> acc(&dst[i])
requires forall i int :: { src[i] } 0 <= i && i < len(src) ==> acc(&src[i], p)
ensures len(res) == len(dst) + len(src)
ensures forall i int :: { res[i] } 0 <= i && i < len(res) ==> acc(&res[i])
ensures forall i int :: { src[i] } 0 <= i && i < len(src) ==> acc(&src[i], p)
ensures forall i int :: { res[i] } 0 <= i && i < len(dst) ==> res[i] == old(dst[i])
ensures forall i int :: { res[i] } len(dst) <= i && i < len(res) ==> res[i] == src[i - len(dst)]
func append(dst []Type, src ...Type, ghost p perm) (res []int)

requires 0 < p && p < 1
requires forall i int :: { dst[i] } (0 <= i && i < len(dst)) ==> acc(&dst[i], 1-p)
requires forall i int :: { src[i] } (0 <= i && i < len(src)) ==> acc(&src[i], p)
requires forall i int :: { dst[i] } (0 <= i && i < len(dst) && (forall j int :: 0 <= j && j < len(src) ==> &dst[i] != &src[j])) ==> acc(&dst[i], p)
ensures len(dst) <= len(src) ==> res == len(dst)
ensures len(src) < len(dst) ==> res == len(src)
ensures forall i int :: { dst[i] } 0 <= i && i < len(dst) ==> acc(&dst[i], 1-p)
ensures forall i int :: { src[i] } 0 <= i && i < len(src) ==> acc(&src[i], p)
ensures forall i int :: { dst[i] } (0 <= i && i < len(dst) && (forall j int :: 0 <= j && j < len(src) ==> &dst[i] != &src[j])) ==> acc(&dst[i], p)
ensures forall i int :: { dst[i] } (0 <= i && i < len(src) && i < len(dst)) ==> dst[i] == old(src[i])
ensures forall i int :: { dst[i] } (len(src) <= i && i < len(dst)) ==> dst[i] == old(dst[i])
func copy(dst, src []Type, ghost p perm) (res int)
```

**UPDATE**: changed the description to reflect the most recent version of the `copy` function

### CODE REVIEW

**Comment by ArquintL** [comment]
I guess you forgot to mention the range of `j` in the description, right?
```
requires forall i int :: (0 <= i && i < len(dst) && forall j int :: &dst[i] != &src[j]) ==> acc(&dst[i], p)
```

File: src/main/scala/viper/gobra/frontend/info/base/BuiltInMemberTag.scala [hunk 1]
```diff
@@ -6,21 +6,13 @@
...
-import viper.gobra.frontend.Config
-import viper.gobra.frontend.info.base.Type.{AbstractType, AssertionT, ChannelModus, ChannelT, FunctionT, IntT, PredT, Type, VoidType}
-import viper.gobra.frontend.info.implementation.typing.ghost.separation.GhostType
-import viper.gobra.util.TypeBounds.UnboundedInteger
-import viper.gobra.util.Violation
-
 
 /**
   * Module to add built-in functions, methods, fpredicates, and mpredicates to Gobra.
   * Two steps have to be performed for adding a new built-in member:
```
Comment (ArquintL) [inline src/main/scala/viper/gobra/frontend/info/base/BuiltInMemberTag.scala:12]:
```suggestion
  * Four steps have to be performed for adding a new built-in member:
```

File: src/main/scala/viper/gobra/frontend/info/base/BuiltInMemberTag.scala [hunk 2]
```diff
@@ -6,21 +6,13 @@
...
   * (1) add a tag representing that built-in member,
   * (2) add it to `builtInMembers()`, and
+  * (3) add type-checking code in file `BuiltInMemberTyping.scala`
   * (3) add an additional case to BuiltInMembersImpl that maps that built-in member (the tag and its specific use) to
```
Comment (ArquintL) [inline src/main/scala/viper/gobra/frontend/info/base/BuiltInMemberTag.scala:16]:
```suggestion
  * (4) add an additional case to BuiltInMembersImpl that maps that built-in member (the tag and its specific use) to
```

File: src/main/scala/viper/gobra/translator/implementations/translator/BuiltInMembersImpl.scala [hunk 3]
```diff
@@ -381,6 +418,152 @@ class BuiltInMembersImpl extends BuiltInMembers {
           in.Access(closedInst, in.FullPerm(src))(src)
         )
         in.Function(x.name, args, Vector(), pres, posts, None)(src)
+
+      case (AppendFunctionTag, Vector(sliceT: in.SliceT, _: in.SliceT)) =>
+        /**
+          * requires forall i int :: 0 <= i && i < len(s) ==> acc(&s[i])
+          * requires forall i int :: 0 <= i && i < len(stuff) ==> acc(&stuff[i], _)
```
Comment (ArquintL) [inline src/main/scala/viper/gobra/translator/implementations/translator/BuiltInMembersImpl.scala:425]: why do you use wildcard amounts? Wouldn't it be better to add another ghost argument such that full permissions to the variadic arguments could be recovered?

**Review by ArquintL — APPROVED** [review]
Looks good, a few minor remarks. It's impressive how few changes very necessary to better integrate the built-ins to the type checker

**Comment by jcp19** [comment]
> I guess you forgot to mention the range of `j` in the description, right?
> 
> ```
> requires forall i int :: (0 <= i && i < len(dst) && forall j int :: &dst[i] != &src[j]) ==> acc(&dst[i], p)
> ```

Thanks for the observation, I did forget to update the description. That is now fixed. This does not impact the code.

**Review by ArquintL — APPROVED** [review]

**Comment by jcp19** [comment]
I did not improve the contract for `append` yet. I'm concerned with the performance of the `copy` implementation, it seems that it is quite easy to make verification diverge in the presence of calls to `copy`

File: src/main/scala/viper/gobra/translator/implementations/translator/BuiltInMembersImpl.scala [hunk 4]
```diff
@@ -385,6 +423,207 @@ class BuiltInMembersImpl extends BuiltInMembers {
           in.Access(closedInst, in.FullPerm(src))(src)
         )
         in.Function(x.name, args, Vector(), pres, posts, None)(src)
+
+      /* João, 18/08/2021:
+       *  The spec for `append` currently does not allow the first and second non-ghost arguments to be the same. The go
+       *  spec however allows that to happen and the behavior is well-defined for those cases (in particular, the result
+       *  of appending two slices is independent of whether they overlap). I did not change the spec to reflect this change
+       *  at the moment to avoid surprises in the performance of Gobra when checking the VerifiedSCION codebase. I do expect
+       *  to change that at a later point.
```
Comment (ArquintL) [inline src/main/scala/viper/gobra/translator/implementations/translator/BuiltInMembersImpl.scala:432]: One idea to hopefully better control performance could be an `overlap` ghost flag that turns "enables" / "disables" one set of spec or the other

**Review by ArquintL — APPROVED** [review]
Looks good, I've only looked at the changed files since my last review

File: src/main/scala/viper/gobra/translator/implementations/translator/BuiltInMembersImpl.scala (hunk 4 above)
Comment (jcp19) [reply src/main/scala/viper/gobra/translator/implementations/translator/BuiltInMembersImpl.scala:432]: I may try that in a future PR. I think we should close this one now, given that it has been open for quite a while.

**Pull request merged**